### PR TITLE
fix(cdk): new autofocus behavior in ios

### LINKED ISSUE #1634: 🐞 - ios autofocus in dialog
### Which @taiga-ui/* package(s) are the source of the bug?

cdk

### Please provide a link to a minimal reproduction of the bug

_No response_

### Description

Dialog "jumps" when tuiAutoFocus is set on input on ios

https://user-images.githubusercontent.com/50827860/163393258-bd7eae0c-fb58-4b75-afbe-84810fa9c506.mov

s

### Angular version

11.2.11

### Taiga UI version

2.41.1

### Which browsers have you used?

- [X] Chrome
- [ ] Firefox
- [X] Safari
- [ ] Edge

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [ ] Linux
- [X] iOS
- [ ] Android

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #1634

https://user-images.githubusercontent.com/12021443/170274403-290e8555-ead6-4322-8cff-ec2d2ef1f06d.MP4

## What is the new behavior?

https://user-images.githubusercontent.com/12021443/170274086-8c4439b4-e24c-40c1-bee1-a9471e95db75.MP4

Example when we change `TUI_ANIMATIONS_DURATION`

```ts
{
        provide: TUI_ANIMATIONS_DURATION,
        useFactory: () => (inject(TUI_IS_CYPRESS) ? 0 : 2000),
    }
```

https://user-images.githubusercontent.com/12021443/170280325-e3045f14-a092-44a6-aeab-afbb3313044e.MP4

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts [hunk 1]
```diff
@@ -73,6 +86,47 @@ export class TuiIosAutofocusHandler extends AbstractTuiAutofocusHandler {
...
+        const defaultAnimationTime =
+            parseFloat(
+                this.windowRef
+                    .getComputedStyle(this.element)
+                    .getPropertyValue('--tui-duration'),
+            ) || 0;
+
+        return this.minTimeOpeningVirtualKeyboard > defaultAnimationTime
+            ? this.minTimeOpeningVirtualKeyboard
+            : defaultAnimationTime;
```
Comment (nsbarsukov) [inline projects/cdk/directives/auto-focus/handlers/ios.handler.ts:130]:
```ts
Math.max(this.minTimeOpeningVirtualKeyboard, defaultAnimationTime);
```

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts [hunk 2]
```diff
@@ -73,6 +86,47 @@ export class TuiIosAutofocusHandler extends AbstractTuiAutofocusHandler {
...
+
+        return fakeInput;
+    }
+
+    private getVisualViewportHeight(): number {
+        /**
+         * Since Chrome 61+, Safari 13+ in Apple/iOS
+         * https://developer.mozilla.org/en-US/docs/Web/API/VisualViewport
+         */
+        type IosWindow = Window & {visualViewport: {height: number}};
```
Comment (nsbarsukov) [inline projects/cdk/directives/auto-focus/handlers/ios.handler.ts:113]: let's make private method with type guard.
Something like:

```ts
private checkHasVisualViewport(
    windowRef: Window
): windowRef is Window & {visualViewport: {height: number}} {
    return 'visualViewport' in this.windowRef;
}
```

**Review by nsbarsukov — CHANGES_REQUESTED** [review]

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts (hunk 1 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/ios.handler.ts]: fixed

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts (hunk 2 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/ios.handler.ts]: fixed

**Review by vladimirpotekhin — APPROVED** [review]

**Review by nsbarsukov — APPROVED** [review]

File: projects/core/styles/markup/tui-mobile-only.less [hunk 3]
```diff
@@ -3,3 +3,13 @@
         display: none !important;
     }
 }
+
+@media @mobile-m {
```
Comment (waterplea) [inline projects/core/styles/markup/tui-mobile-only.less:7]: This isn't the place for this

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts [hunk 4]
```diff
@@ -37,42 +42,78 @@ export class TuiIosAutofocusHandler extends AbstractTuiAutofocusHandler {
...
+
+        fakeInput.style.height = px(this.element.clientHeight);
+        fakeInput.style.width = px(this.element.clientWidth / 2);
+        fakeInput.style.position = 'fixed';
+        fakeInput.style.opacity = '0';
+        fakeInput.style.fontSize = px(16); // disable possible auto zoom
+        fakeInput.readOnly = true; // prevent keyboard for fake input
+
+        // @note: emulate position cursor before focus to real textfield element
+        fakeInput.style.top = px(this.element.getBoundingClientRect().top);
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/handlers/ios.handler.ts:93]: You shouldn't request client rect twice, it causes a reflow. Also clientHeight and clientWidth, just get them from the same clientRect - request it once and store in a constant.

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts [hunk 5]
```diff
@@ -37,42 +42,78 @@ export class TuiIosAutofocusHandler extends AbstractTuiAutofocusHandler {
...
-         * the browser automatically focuses on the invisible element.
-         * And then, after focus is triggered, we try to focus on target element, and if we managed to refocus,
-         * then we try to focus again on an invisible element, so that the keyboard slowly appears.
-         * This ping pong allows the keyboard to not overlap the modal window.
-         */
         fakeInput.addEventListener('blur', blurHandler, {once: true});
         fakeInput.addEventListener('focus', focusHandler);
 
-        this.element.parentElement?.appendChild(fakeInput);
+        if (this.insideDialog()) {
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/handlers/ios.handler.ts:68]: Is this branching really important? Maybe just always append it to body?

File: projects/core/styles/markup/tui-mobile-only.less [hunk 6]
```diff
@@ -3,3 +3,13 @@
...
+    html,
+    body {
+        // only for iOS devices
+        @supports (-webkit-touch-callout: none) {
+            overflow: auto;
```
Comment (nsbarsukov) [inline projects/core/styles/markup/tui-mobile-only.less:12]: can we somehow encapsulate this logic inside `auto-focus`-directive ?
I am afraid that we can forget later that this line is closely connected with this directive(

(or add comment, at least?)

Comment (splincode) [reply projects/core/styles/markup/tui-mobile-only.less]: fixed

File: projects/cdk/directives/auto-focus/handlers/ios.handler.ts (hunk 5 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/ios.handler.ts]: fixed

File: projects/core/styles/markup/tui-mobile-only.less (hunk 3 above)
Comment (splincode) [reply projects/core/styles/markup/tui-mobile-only.less]: fixed

**Review by nsbarsukov — APPROVED** [review]

**Pull request merged**